qiskit_ibm/runtime/utils.py

```diff
@@ -193,4 +198,17 @@ def object_hook(self, obj: Any) -> Any:
                 return _decode_and_deserialize(obj_val, scipy.sparse.load_npz, False)
             if obj_type == 'to_json':
                 return obj_val
+        if isinstance(obj, dict) and self.USE_INT_KEYS:
+            keys_to_add: List[int] = []
+            # Cast integer keys disguised as strings back to integer keys
+            for key in obj.keys():
+                try:
+                    keys_to_add.insert(0, int(key))
+                except ValueError:
+                    pass
```

Comment:
This means the users would need to know to set `RuntimeDecoder.USE_INT_KEYS` to `True` if they want to convert the key back to integer, and there's the risk of accidentally converting the wrong keys (i.e. strings that look like integers but meant to remain as string). How about having the encoder checking whether the key is indeed an integer and save that information in the return dictionary?

Reply:
Updated! `use_int_keys` is now stored in the serialized json object abstracted from the user.